### PR TITLE
[flutter_tools] pretty print hot reload rejection error

### LINKED ISSUE #64027: Hot reload failure error message is ugly
Currently if you remove a field from a const class and try to hot reload you get:

```
Hot reload received invalid response: {type: ReloadReport, success: false, notices: [{type: ReasonForCancelling, class: {type: @Class, fixedId: true, id:
classes/3988, name: ItemGeometry}, message: Const class cannot remove fields: Library:'package:plexiglass/rendering/items/common.dart' Class: ItemGeometry}]}
```

This is not a pretty message. We should pretty print it.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

If the vm of an attached device rejects a hot reload, pretty print the reason. Suggest a hot restart so that users are aware that they do not have to detach and rebuild. Also resets the last compilation time, so a subsequent restart would still apply the last change

Adds an integration test for the const field removal.

Fixes https://github.com/flutter/flutter/issues/64027

### CODE REVIEW

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -885,7 +868,8 @@ class HotRunner extends ResidentRunner {
             nullSafety: usageNullSafety,
             fastReassemble: null,
           ).send();
-          return OperationResult(1, 'Reload rejected');
+          final ReloadReportContents contents = ReloadReportContents.fromReloadReport(reloadReport);
+          return OperationResult(1, 'Reload rejected: ${contents.notices.join("\n")}');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/run_hot.dart:872]: including the additional context in the operation result is required to surface this information in IDEs

File: packages/flutter_tools/test/integration.shard/test_driver.dart [hunk 2]
```diff
@@ -588,19 +588,6 @@ class FlutterRunTestDriver extends FlutterTestDriver {
     );
   }
 
-  Future<void> reloadMethod({ String libraryId, String classId }) async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/test_driver.dart:591]: unused code

File: packages/flutter_tools/test/integration.shard/test_utils.dart [hunk 3]
```diff
@@ -55,6 +55,7 @@ Future<void> getPackages(String folder) async {
     fileSystem.path.join(getFlutterRoot(), 'bin', 'flutter'),
     'pub',
     'get',
+    '--offline',
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/test_utils.dart:58]: might save some network time, these should only depend on flutter itself

File: packages/flutter_tools/lib/src/devfs.dart [hunk 4]
```diff
@@ -440,6 +441,20 @@ class DevFS {
...
+  /// Mark the [lastCompiled] time to the previous successful compile.
+  ///
+  /// Sometimes a hot reload will be rejected by the VM due to a change in the
+  /// structure of the code not supporting the hot reload. In these cases,
+  /// the best resolution is a hot restart. However, the resident runner
+  /// will not recognize this file as having been changed since the delta
+  /// will already have been accepted. Instead, reset the compile time so
+  /// that the last updated files are included in subsequent compilations until
+  /// a reload is accepted.
+  void resetLastCompiled() {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:453]: Still needs a test

File: packages/flutter_tools/test/general.shard/resident_runner_test.dart [hunk 5]
```diff
@@ -189,21 +189,6 @@ void main() {
     when(mockFlutterDevice.vmService).thenAnswer((Invocation invocation) {
       return fakeVmServiceHost?.vmService;
     });
-    when(mockFlutterDevice.reloadSources(any, pause: anyNamed('pause'))).thenAnswer((Invocation invocation) async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/resident_runner_test.dart:192]: I delete some mocks...

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 6]
```diff
@@ -1307,3 +1333,57 @@ class ProjectFileInvalidator {
     );
   }
 }
+
+/// Additional serialization logic for a hot reload response.
+class ReloadReportContents {
+  factory ReloadReportContents.fromReloadReport(vm_service.ReloadReport report) {
+    final List<ReasonForCancelling> reasons = <ReasonForCancelling>[];
+    final Object notices = report.json['notices'];
+    if (notices is List<dynamic>) {
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart:1342]: Done!

**Comment by jonahwilliams** [comment]
Google testing passed according to the dashboard.

**Pull request merged**